gapis/api/vulkan/overdraw.go

```diff
@@ -0,0 +1,1322 @@
+	secCmdBuffers := map[VkCommandBuffer]VkCommandBuffer{}
+
+	rpEnded := false
+	for i := 0; i < bInfo.CommandReferences().Len(); i++ {
+		cr := bInfo.CommandReferences().Get(uint32(i))
+		args := GetCommandArgs(ctx, cr, st)
+		if uint64(i) >= rpStartIdx && !rpEnded {
+			switch ar := args.(type) {
+			case VkCmdBeginRenderPassArgsʳ:
+				newArgs := ar.Clone(a, api.CloneContext{})
+				newArgs.SetRenderPass(renderInfo.renderPass)
+				newArgs.SetFramebuffer(renderInfo.framebuffer)
+
+				clearCount := uint32(newArgs.ClearValues().Len())
+				newClear := NewU32ː4ᵃ(a)
+
+				if renderInfo.depthIdx != ^uint32(0) {
+					newClear.Set(0, newArgs.
+						ClearValues().
+						Get(renderInfo.depthIdx).
+						Color().
+						Uint32().
+						Get(0))
+				}
+				// 0 initialize the stencil buffer
+				newArgs.ClearValues().Add(clearCount,
+					// Use VkClearColorValue instead of
+					// VkClearDepthValue because it doesn't
+					// seem like the union is set up in the
+					// API DSL
+					NewVkClearValue(a, NewVkClearColorValue(a,
+						newClear)))
+				args = newArgs
+			case VkCmdEndRenderPassArgsʳ:
+				rpEnded = true
+			case VkCmdBindPipelineArgsʳ:
+				newArgs := ar
+				if ar.PipelineBindPoint() ==
+					VkPipelineBindPoint_VK_PIPELINE_BIND_POINT_GRAPHICS {
+					newArgs = ar.Clone(a, api.CloneContext{})
+
+					pipe := ar.Pipeline()
+					newPipe, ok := pipelines[pipe]
+					if !ok {
+						var err error
+						newPipe, err = s.createGraphicsPipeline(ctx, cb, gs, st,
+							a, device, pipe, renderInfo.renderPass, alloc,
+							addCleanup, out)
+						if err != nil {
+							return 0, err
+						}
+						pipelines[pipe] = newPipe
+					}
+					newArgs.SetPipeline(newPipe)
+				}
+				args = newArgs
+			case VkCmdExecuteCommandsArgsʳ:
+				newArgs := ar
+				for i := uint32(0); i < uint32(ar.CommandBuffers().Len()); i++ {
+					cmdbuf := ar.CommandBuffers().Get(i)
+					newCmdbuf, ok := secCmdBuffers[cmdbuf]
+					if !ok {
+						var err error
+						newCmdbuf, err = s.createCommandBuffer(ctx,
+							cb, gs, st, a, cmdbuf, renderInfo,
+							0, alloc, addCleanup, out)
+						if err != nil {
+							return 0, err
+						}
+						secCmdBuffers[cmdbuf] = newCmdbuf
+					}
+					newArgs.CommandBuffers().Add(i, newCmdbuf)
+				}
+				args = newArgs
+			}
+		}
+		cleanup, cmd, _ := AddCommand(ctx, cb, newCmdBuffer, gs,
+			gs, args)
+
+		writeEach(ctx, out, cmd)
+		cleanup()
+	}
+	writeEach(ctx, out,
+		cb.VkEndCommandBuffer(newCmdBuffer, VkResult_VK_SUCCESS))
+
+	return newCmdBuffer, nil
+}
+
+func (s *stencilOverdraw) rewriteQueueSubmit(ctx context.Context,
+	cb CommandBuilder,
+	gs *api.GlobalState,
+	st *State,
+	a arena.Arena,
+	submit *VkQueueSubmit,
+	rpBeginArgs VkCmdBeginRenderPassArgsʳ,
+	rpBeginIdx api.SubCmdIdx,
+	cmdId api.CmdID,
+	alloc func(v ...interface{}) api.AllocResult,
+	addCleanup func(func()),
+	out transform.Writer,
+) (stencilImage, error) {
+	// TODO: check if we're allowed to modify the command directly, since
```

Comment:
You shouldn't modify the commands. This can cause problems later on, since these are shared data-structures. (If golang has a CONST keyword we would be using it here).